### PR TITLE
Fixes #9618 - make identifier uniq per host

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

**Comment by ares** [comment]
replaces #2218

**Comment by dLobatog** [comment]
@ares tests failing

**Comment by ares** [comment]
@eLobato thanks, I forgot to ignore empty identifiers as they are treated as separate validation, should be fixed now

File: app/models/host/base.rb [hunk 1]
```diff
@@ -370,5 +371,23 @@ def check_provision_interface
         errors.add :interfaces, _("managed host must have one provision interface")
       end
     end
+
+    # we can't use standard unique validation on interface since we can't properly handle :scope => :host_id
+    # for new hosts host_id does not exist at that moment, validation would work only for persisted records
+    def uniq_interfaces_identifiers
+      success = true
+      identifiers = []
```
Comment (unorthodoxgeek) [inline app/models/host/base.rb:379]: I don't like using local variables when they also mean a method name on self. can you rename this local variable for sake of less confusion?

File: test/unit/host_test.rb [hunk 2]
```diff
@@ -1123,6 +1123,19 @@ def teardown
...
 
+    test "host can't have more interfaces with same identifier" do
+      host = FactoryGirl.build(:host, :managed)
+      host.primary_interface.identifier = 'eth0'
+      nic = host.interfaces.build(:identifier => 'eth0')
+      refute host.valid?
+      assert_present nic.errors[:identifier]
+      assert_present host.errors[:interfaces]
+      nic.identifier = 'eth1'
+      host.valid?
```
Comment (unorthodoxgeek) [inline test/unit/host_test.rb:1134]: valid? is logically a question, and makes little sense in this context...
can you put this into context by doing

``` ruby
assert host.invalid?
```

in rails 4 you can host.validate! which does make sense without assertion, but we're not there yet.

File: app/models/host/base.rb (hunk 1 above)
Comment (ares) [inline app/models/host/base.rb:379]: host does not have `identifiers` method AFAIK, interfaces have one similar, called `identifier`... anyway I can't think of better name for variable that holds interface identifiers

File: test/unit/host_test.rb (hunk 2 above)
Comment (ares) [inline test/unit/host_test.rb:1134]: this is on purpose, this test does not know whether the host will be valid? or not, there might be other validations added later, we just need to trigger all validations and then make sure that there are no errors for identifier and interfaces

note that in few lines above `refute` is also intentional, as we are pretty sure that we're making host invalid by setting bad identifier

**Comment by ares** [comment]
@unorthodoxgeek any other concerns? do you consider your concerns explained?

File: app/models/host/base.rb (hunk 1 above)
Comment (unorthodoxgeek) [inline app/models/host/base.rb:379]: hehe, I think I was really tired when I reviewed this.
I read "identifiers" and "interfaces" as the same word.
:facepalm:

**Comment by unorthodoxgeek** [comment]
@ares  - all good
I think this can be merged :+1:

File: test/unit/host_test.rb (hunk 2 above)
Comment (unorthodoxgeek) [inline test/unit/host_test.rb:1134]: ok, I get it.
in rails 4 this should be changed to host.validate! so it won't be an open question, it mainly disturbs me grammatically :smile:

**Comment by ares** [comment]
ok, removed need rereview label, did you also tested it locally?

**Comment by unorthodoxgeek** [comment]
no, I didn't, what's the use case?

**Comment by ares** [comment]
To make identifier uniq per host, so you can't create a single host with two "eth0" interfaces. During reviews, please make sure you go through all items in http://projects.theforeman.org/projects/foreman/wiki/Reviewing_patches this should probably fall under the first one.

File: app/models/host/base.rb [hunk 3]
```diff
@@ -370,5 +371,23 @@ def check_provision_interface
...
+      self.interfaces.each do |interface|
+        next if interface.identifier.blank?
+        if identifiers.include?(interface.identifier)
```
Comment (ShimShtein) [inline app/models/host/base.rb:382]: Why not 

``` ruby
groups = self.interfaces.group_by { |interface| interface.identifier }
dups = groups.find_all { |group| group.length > 1 }
```

Comment (unorthodoxgeek) [inline app/models/host/base.rb:382]: that's too smart :stuck_out_tongue_winking_eye: 
in code, follow the keep it simple principle...

Comment (ShimShtein) [inline app/models/host/base.rb:382]: The KISS principle is what led me to think that two instruction solution is better than seven :stuck_out_tongue:

Comment (unorthodoxgeek) [inline app/models/host/base.rb:382]: it's not only about number of lines, it's about the amount of brain power required to read those line...

Comment (ares) [inline app/models/host/base.rb:382]: usually I like less imperative style, but even with this, you'd have to set success = false if dups.present? etc... I don't think it's much better in this case

**Comment by ares** [comment]
anything I could do to make this merged? it's very easy to test, just try to create host with two interfaces with same identifier, you should see validation error

**Comment by dLobatog** [comment]
Merged as f2d5f95539b5070f5c7aba37196c45346140357d, thanks @ares

**Pull request closed without merging**